### PR TITLE
Refactor CMake

### PR DESCRIPTION
**Pull request by jerryct**

### CODE REVIEW

**Comment by gjasny** [comment]
Hello,

thank you for your contribution to prometheus-cpp. While I definitely welcome the consistent formatting changes (how did you create them? manually?) some of the other changes break the CMake 3.0 compatibility. Could you please drop those?

Thanks,
Gregor

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
 project(prometheus-cpp)
-enable_testing()
+
 include(GNUInstallDirs)
 
-set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_CURRENT_LIST_DIR}/cmake)
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
 
-if(NOT DEFINED CMAKE_CXX_STANDARD AND UNIX)
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
```
Comment (gjasny) [inline CMakeLists.txt]: That only works for CMake ~3.4.

File: core/CMakeLists.txt [hunk 2]
```diff
@@ -15,21 +15,31 @@ add_library(prometheus-cpp-core
...
+  PRIVATE
+    ${CMAKE_THREAD_LIBS_INIT}
+    $<$<AND:$<BOOL:UNIX>,$<NOT:$<BOOL:APPLE>>>:rt>
+)
 
-target_include_directories(prometheus-cpp-core PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
+target_include_directories(prometheus-cpp-core
+  PUBLIC
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+    $<INSTALL_INTERFACE:include>
```
Comment (gjasny) [inline core/CMakeLists.txt]: The `INSTALL_INTERFACE` here is wrong and is duplicated by line 29/36 `INCLUDES DESTINATION`

File: core/CMakeLists.txt [hunk 3]
```diff
@@ -15,21 +15,31 @@ add_library(prometheus-cpp-core
...
+)
 
-install(TARGETS prometheus-cpp-core EXPORT prometheus-cpp-targets
+install(
+  TARGETS prometheus-cpp-core
+  EXPORT prometheus-cpp-targets
```
Comment (gjasny) [inline core/CMakeLists.txt]: For consistency `${PROJECT_NAME}-targets`? But I'd find the `prometheus-cpp-targets` more readable.

File: pull/CMakeLists.txt [hunk 4]
```diff
@@ -12,30 +12,42 @@ add_library(prometheus-cpp-pull
...
+    ${CMAKE_THREAD_LIBS_INIT}
+    $<$<AND:$<BOOL:UNIX>,$<NOT:$<BOOL:APPLE>>>:rt>
+)
 
-target_include_directories(prometheus-cpp-pull PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
-target_include_directories(prometheus-cpp-pull PRIVATE ${CIVETWEB_INCLUDE_DIR})
+target_include_directories(prometheus-cpp-pull
+  PUBLIC
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+    $<INSTALL_INTERFACE:include>
```
Comment (gjasny) [inline pull/CMakeLists.txt]: please drop

File: push/CMakeLists.txt [hunk 5]
```diff
@@ -11,24 +11,37 @@ add_library(prometheus-cpp-push
...
+    ${CURL_LIBRARIES}
+    $<$<AND:$<BOOL:UNIX>,$<NOT:$<BOOL:APPLE>>>:rt>
+)
 
-target_include_directories(prometheus-cpp-push PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
-target_include_directories(prometheus-cpp-push PRIVATE ${CPR_INCLUDE_DIR} ${CURL_INCLUDE_DIRS})
+target_include_directories(prometheus-cpp-push
+  PUBLIC
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+    $<INSTALL_INTERFACE:include>
```
Comment (gjasny) [inline push/CMakeLists.txt]: please drop

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,17 +1,17 @@
 cmake_minimum_required(VERSION 2.8.12.2 FATAL_ERROR)
+
 if(POLICY CMP0063)
```
Comment (gjasny) [inline CMakeLists.txt]: Please rebase onto master. Some of the CMake stuff has changed there.

File: CMakeLists.txt (hunk 1 above)
Comment (jerryct) [reply CMakeLists.txt]: It is available with 3.1 :) (https://cmake.org/cmake/help/v3.1/variable/CMAKE_CXX_STANDARD.html?highlight=cmake_cxx_standard). Why was 3.0 chosen? Is it possible to raise the version? (most Linux distributions ship with 3.5 or higher and on other platforms which do not have a package manager the user has to install CMake anyway manually)

Comment (gjasny) [reply CMakeLists.txt]: It might make sense to raise the limit to 3.5. Both Ubuntu trusty and RHEL 6 EPEL provide a CMake 3.5 (or later package)

**Comment by jerryct** [comment]
I rebased it.

File: core/CMakeLists.txt [hunk 7]
```diff
@@ -1,4 +1,5 @@
-add_library(prometheus-cpp-core
+
+add_library(core
```
Comment (gjasny) [inline core/CMakeLists.txt]: The problem with renaming the targets from `prometheus-cpp-core` to `core` is that also the library files get the short name:
```
./core/libcore.a
./pull/libpull.a
./push/libpush.a
```
I'd prefer if you restore the long names and set the `EXPORT_NAME`, again.

Thanks,
Gregor

Comment (jerryct) [reply core/CMakeLists.txt]: Hi. Would you be ok to make your request more explicit, i.e., explicitly stating to give the library a specific name:

https://github.com/jupp0r/prometheus-cpp/pull/166/files#diff-5f4a237561b8048385b3ba09ef74fc96R30

while naming the targets internally and externally the same.

**Pull request merged**